openprescribing/frontend/ghost_branded_generics.py

```diff
@@ -0,0 +1,261 @@
+    prescribing = get_prescribing_for_orgs(db, bnf_codes, date, org_type, org_ids)
+    results = []
+    bnf_codes_used = set()
+    for org_id, bnf_code, quantities, net_costs in prescribing:
+        tariff_price = prices[bnf_code]
+        tariff_costs = quantities * tariff_price
+        possible_savings = net_costs - tariff_costs
+        savings_above_threshold = (
+            numpy.absolute(possible_savings) >= MIN_GHOST_GENERIC_DELTA
+        )
+        total_savings = possible_savings.sum(where=savings_above_threshold)
+        if total_savings != 0:
+            bnf_codes_used.add(bnf_code)
+            total_net_cost = net_costs.sum()
+            total_quantity = quantities.sum()
+            results.append(
+                {
+                    "date": date,
+                    "org_type": org_type,
+                    "org_id": org_id,
+                    "bnf_code": bnf_code,
+                    "median_ppu": tariff_price / 100,
+                    "price_per_unit": total_net_cost / total_quantity / 100,
+                    "quantity": total_quantity,
+                    "possible_savings": total_savings / 100,
+                }
+            )
+    names = Presentation.names_for_bnf_codes(bnf_codes_used)
+    for result in results:
+        result["product_name"] = names.get(result["bnf_code"], "unknown")
+    results.sort(key=lambda i: i["possible_savings"], reverse=True)
+    return results
+
+
+def get_total_ghost_branded_generic_spending(date, org_type, org_id):
+    """
+    Get the total spend on generics (by this org and in this month) over and
+    above the price set by the Drug Tariff
+    """
+    db = get_db()
+    practice_spending = get_total_ghost_branded_generic_spending_per_practice(
+        db, date, PRESENTATIONS_TO_IGNORE, MIN_GHOST_GENERIC_DELTA
+    )
+    group_by_org = get_row_grouper(org_type)
+    return group_by_org.sum_one_group(practice_spending, org_id)[0] / 100
+
+
+@memoize()
+def get_total_ghost_branded_generic_spending_per_practice(
+    db, date, presentations_to_ignore, min_delta
+):
+    """
+    Get the total spend on generics by all practices in this month over and
+    above the price set by the Drug Tariff.
+
+    This gives us a single, cacheable matrix from which we can calculate totals
+    for any organisation.
+    """
+    prices = get_inferred_tariff_prices(db, date, presentations_to_ignore)
+    bnf_codes = list(prices.keys())
+    totals = None
+    for bnf_code, quantities, net_costs in get_prescribing(db, bnf_codes, date):
+        if not isinstance(quantities, numpy.ndarray):
+            quantities = quantities.toarray()
+        if not isinstance(net_costs, numpy.ndarray):
+            net_costs = net_costs.toarray()
+        tariff_price = prices[bnf_code]
+        tariff_costs = quantities * tariff_price
+        possible_savings = net_costs - tariff_costs
+        savings_above_threshold = numpy.absolute(possible_savings) >= min_delta
```

Comment:
Sorry if I've misread (I'm really only skimming here) but I remember a discussion where we agreed only positive savings should be added together?

Reply:
Yes, we did discuss (I'm still not quite sure how negative savings are even possible here). But we decided it would need Brian and Rich's input and they're busy with other things. It's easy to change later and I figured mirroring our existing results as closely as possible was OK for now.